src/persist/src/storage.rs

```diff
@@ -330,6 +330,167 @@ impl From<(&str, &str)> for LockInfo {
     }
 }
 
+/// Abstraction over a key value store that provides a compare-and-set/conditional put primitive
+///
+/// "real world" realizations of this api could be etcd / dynamodb / any sql database with transactions?
+pub trait CasKeyVal {
+    /// Returns a reference to the value corresponding to the key.
+    fn get(&self, key: &str) -> Result<Option<Vec<u8>>, Error>;
+
+    /// List all of the keys in the map that currently have a value.
+    fn list_keys(&self) -> Result<Vec<String>, Error>;
+
+    /// Conditionally insert/update a key-value pair into the map, if the existing
+    /// value matches expected.
+    ///
+    /// Return true if the cas executed correctly, and false if the current value != expected.
+    ///
+    /// TODO: could also have the return value be the current value here. Doing a simpler thing
+    /// for prototyping.
+    fn cas(
+        &mut self,
+        key: &str,
+        value: Option<Vec<u8>>,
+        expected: Option<Vec<u8>>,
+    ) -> Result<bool, Error>;
+}
+
+/// Abstraction over a key value store that supports atomic inserts, and scans
+/// from a monotonically increasing index
+///
+/// "real world" realizations of this api could be kafka / redpanda.
+pub trait AtomicQueue {
```

Comment:
I didn't actually do this in the first commit!

Reply:
I'm a little confused. Possibly we're talking past each other? I was responding here to your comment in slack of "i guess i lied - i did one, and i will do the other one in a bit"

Reply:
sorry sorry - I meant - "I implemented Consensus on top of CasKeyVal, and pushed it to this pr and I have not yet implemented Consensus on top of AtomicQueue" and then I said "I didn't actually do this" where this== "implement Consensus on top of AtomicQueue" because you said you were nerd sniped

Reply:
oh, ha! I said "don't bother doing this" and then I said "nevermind, I'm curious how you were going to do this". by "nerd sniped" I didn't mean I was actually doing anything with it